sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,727 @@
+        centroids_half_l2_norm,             # IN    (n_clusters,)
+        per_sample_pseudo_inertia,          # OUT   (n_samples,)
+        new_centroids_t_private_copies,     # OUT   (n_private_copies, n_features, n_clusters)
+        cluster_sizes_private_copies,       # OUT   (n_private_copies, n_clusters)
+    ):
+    # fmt: on
+        """One full iteration of LLoyd's k-means.
+
+        The kernel is meant to be invoked on a 1D grid spanning the data samples of
+        the training set in parallel.
+
+        Each work-item will assign that sample to its nearest centroid and accumulate
+        the feature values of the data samples into the new version of the centroid
+        for the next iteration. The centroid assignment and centroid update steps are
+        performed in a single fused kernel to avoid introducing a large intermediate
+        label assignment array to be re-read from global memory before performing the
+        centroid update step.
+
+        To avoid atomic contention when performing the centroid updates concurrently
+        for different data samples, we use private copies of the new centroid array in
+        global memory. Those copies are meant to be re-reduced afterwards.
+
+        The distance is the euclidean distance. Note that it is not necessary to
+        compute the exact value to find the closest centroid. Indeed, minimizing
+            |x-y|^2 = |x|^2 - 2<x.y> + |y|^2
+        over y amounts to minimizing
+            (1/2)y^2 - xy .
+        Moreover the value (1/2)y^2 has been pre-computed in the array
+        centroids_half_l2_norm to reduce the overall number of floating point
+        operations in the kernel.
+        """
+        sample_idx = dpex.get_global_id(0)
+        local_work_id = dpex.get_local_id(0)
+
+        # This array in shared memory is used as a sliding array over values of
+        # current_centroids_t. During each iteration in the inner loop, a new one is 
+        # loaded and used by all work items in the work group to compute partial 
+        # results. The array slides over the features in the outer loop, and over the 
+        # samples in the inner loop.
+        centroids_window = dpex.local.array(shape=centroids_window_shape, dtype=dtype)
+
+        # This array in shared memory is used as a sliding array over the centroids.
+        # It contains values of centroids_half_l2_norm for each centroid in the sliding
+        # centroids_window array. It is updated once per iteration in the outer loop.
+        centroids_window_half_l2_norm = dpex.local.array(
+            shape=window_n_centroids, dtype=dtype
+        )
+
+        # In the inner loop each work item accumulates in private memory the
+        # dot product of the sample at the sample_idx relatively to each centroid
+        # in the window.
+        partial_dot_products = dpex.private.array(
+            shape=window_n_centroids, dtype=dtype
+        )
```

Comment:
I think this is more exact. If taken in account, the subsequent occurrences of `partial_dot_products` must also be renamed to `partial_half_squared_differences`.
```suggestion
        # In the inner loop each work item accumulates in private memory a part
        # of the squared difference of the sample at the sample_idx relatively
        # to each centroid in the window, namely: (1/2)y² - xy 
        partial_half_squared_differences = dpex.private.array(
            shape=window_n_centroids, dtype=dtype
        )
```

Reply:
What is accumulated in private memory and in the inner loop really is the dot product, nothing more. The partial half squared differences is what would be called `per sample pseudo inertia` in this code and is computed once the dot product is entierly computed.